lib/cadet/auth/providers/github.ex

```diff
@@ -0,0 +1,76 @@
+defmodule Cadet.Auth.Providers.GitHub do
+  @moduledoc """
+  Provides identity using GitHub OAuth.
+  """
+  alias Cadet.Auth.Provider
+
+  @behaviour Provider
+
+  @type config :: %{client_secret: String.t()}
+
+  @token_url "https://github.com/login/oauth/access_token"
+  @user_api "https://api.github.com/user"
+
+  @spec authorise(config(), Provider.code(), Provider.client_id(), Provider.redirect_uri()) ::
+          {:ok, %{token: Provider.token(), username: String.t()}}
+          | {:error, Provider.error(), String.t()}
+  def authorise(config, code, client_id, redirect_uri) do
+    token_query =
+      URI.encode_query(%{
+        client_id: client_id,
+        client_secret: config.client_secret,
```

Comment:
Should either

1. have a map of client_id => client_secret in the provider config (preferred); or
2. have client_id in the provider config and check that the client_id matches the one in the provider config

LumiNUS is a bit special because NUS ADFS is not configured to use any client secret.

The reason we want to accept multiple client IDs is so that we can use frontends hosted in multiple locations with a single backend. (Each client ID typically has a different (set of) accepted redirect URL(s).)